File: deprecated/pkg/steam/shortcuts.go
package steam

import (
	"fmt"
	"hash/crc32"
	"os"
	"path/filepath"
	"strings"

	"github.com/lobinuxsoft/capydeploy/pkg/protocol"
)

// ShortcutManager handles Steam shortcut operations.
type ShortcutManager struct {
	paths *Paths
}

// NewShortcutManager creates a new ShortcutManager.
func NewShortcutManager() (*ShortcutManager, error) {
	paths, err := NewPaths()
	if err != nil {
		return nil, err
	}
	return &ShortcutManager{paths: paths}, nil
}

// NewShortcutManagerWithPaths creates a ShortcutManager with custom paths.
func NewShortcutManagerWithPaths(paths *Paths) *ShortcutManager {
	return &ShortcutManager{paths: paths}
}

// GetShortcutsPath returns the shortcuts.vdf path for a user.
func (m *ShortcutManager) GetShortcutsPath(userID string) string {
	return m.paths.ShortcutsPath(userID)
}

// GetGridDir returns the grid artwork directory for a user.
func (m *ShortcutManager) GetGridDir(userID string) string {
	return m.paths.GridDir(userID)
}

// EnsureGridDir creates the grid directory if it doesn't exist.
func (m *ShortcutManager) EnsureGridDir(userID string) error {
	return m.paths.EnsureGridDir(userID)
}

// GenerateAppID generates a Steam shortcut app ID from executable path and name.
// This matches Steam's algorithm for non-Steam game shortcuts.
func GenerateAppID(exe, name string) uint32 {
	// Steam's algorithm: CRC32 of (exe + name) | 0x80000000
	key := exe + name
	crc := crc32.ChecksumIEEE([]byte(key))
	// Top bit set to mark as shortcut, bottom bit set
	return (crc | 0x80000000) | 0x02000000
}

// ConvertToShortcutInfo converts a protocol.ShortcutConfig to protocol.ShortcutInfo.
func ConvertToShortcutInfo(cfg protocol.ShortcutConfig) protocol.ShortcutInfo {
	return protocol.ShortcutInfo{
		AppID:         GenerateAppID(cfg.Exe, cfg.Name),
		Name:          cfg.Name,
		Exe:           cfg.Exe,
		StartDir:      cfg.StartDir,
		LaunchOptions: cfg.LaunchOptions,
		Tags:          cfg.Tags,
	}
}

// ArtworkPaths returns all artwork paths for a shortcut.
func (m *ShortcutManager) ArtworkPaths(userID string, appID uint32) map[ArtworkType]string {
	return map[ArtworkType]string{
		ArtworkGrid:     m.paths.ArtworkPath(userID, appID, ArtworkGrid, "png"),
		ArtworkHero:     m.paths.ArtworkPath(userID, appID, ArtworkHero, "png"),
		ArtworkLogo:     m.paths.ArtworkPath(userID, appID, ArtworkLogo, "png"),
		ArtworkIcon:     m.paths.ArtworkPath(userID, appID, ArtworkIcon, "png"),
		ArtworkPortrait: m.paths.ArtworkPath(userID, appID, ArtworkPortrait, "png"),
	}
}

// FindExistingArtwork finds existing artwork files for an appID.
func (m *ShortcutManager) FindExistingArtwork(userID string, appID uint32) (map[ArtworkType]string, error) {
	gridDir := m.paths.GridDir(userID)
	result := make(map[ArtworkType]string)

	extensions := []string{"png", "jpg", "jpeg", "ico"}
	artTypes := map[ArtworkType]string{
		ArtworkGrid:     fmt.Sprintf("%d", appID),
		ArtworkHero:     fmt.Sprintf("%d_hero", appID),
		ArtworkLogo:     fmt.Sprintf("%d_logo", appID),
		ArtworkIcon:     fmt.Sprintf("%d_icon", appID),
		ArtworkPortrait: fmt.Sprintf("%dp", appID),
	}

	for artType, baseName := range artTypes {
		for _, ext := range extensions {
			path := filepath.Join(gridDir, baseName+"."+ext)
			if _, err := os.Stat(path); err == nil {
				result[artType] = path
				break
			}
		}
	}

	return result, nil
}

// SaveArtwork saves artwork data to the appropriate path.
func (m *ShortcutManager) SaveArtwork(userID string, appID uint32, artType ArtworkType, data []byte, ext string) error {
	if err := m.EnsureGridDir(userID); err != nil {
		return fmt.Errorf("failed to create grid dir: %w", err)
	}

	// Clean extension
	ext = strings.TrimPrefix(ext, ".")
	if ext == "" {
		ext = "png"
	}

	path := m.paths.ArtworkPath(userID, appID, artType, ext)
	return os.WriteFile(path, data, 0644)
}

// DeleteArtwork removes all artwork for an appID.
func (m *ShortcutManager) DeleteArtwork(userID string, appID uint32) error {
	existing, err := m.FindExistingArtwork(userID, appID)
	if err != nil {
		return err
	}

	for _, path := range existing {
		if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
			return fmt.Errorf("failed to remove %s: %w", path, err)
		}
	}

	return nil
}

File: crates/steam/src/shortcuts.rs
use std::collections::HashMap;
use std::fs;

use capydeploy_protocol::{ShortcutConfig, ShortcutInfo};
use crc32fast::Hasher;

use crate::SteamError;
use crate::paths::{ArtworkType, Paths};

/// Handles Steam shortcut operations and artwork management.
pub struct ShortcutManager {
    paths: Paths,
}

impl ShortcutManager {
    /// Creates a new `ShortcutManager` with auto-detected Steam paths.
    pub fn new() -> Result<Self, SteamError> {
        Ok(Self {
            paths: Paths::new()?,
        })
    }

    /// Creates a `ShortcutManager` with custom paths.
    pub fn with_paths(paths: Paths) -> Self {
        Self { paths }
    }

    /// Returns the shortcuts.vdf path for a user.
    pub fn shortcuts_path(&self, user_id: &str) -> String {
        self.paths
            .shortcuts_path(user_id)
            .to_string_lossy()
            .into_owned()
    }

    /// Returns the grid artwork directory for a user.
    pub fn grid_dir(&self, user_id: &str) -> String {
        self.paths.grid_dir(user_id).to_string_lossy().into_owned()
    }

    /// Creates the grid directory if it doesn't exist.
    pub fn ensure_grid_dir(&self, user_id: &str) -> Result<(), SteamError> {
        self.paths.ensure_grid_dir(user_id)
    }

    /// Returns all artwork paths for a shortcut.
    pub fn artwork_paths(&self, user_id: &str, app_id: u32) -> HashMap<ArtworkType, String> {
        let mut map = HashMap::new();
        for &art_type in ArtworkType::all() {
            let path = self.paths.artwork_path(user_id, app_id, art_type, "png");
            map.insert(art_type, path.to_string_lossy().into_owned());
        }
        map
    }

    /// Finds existing artwork files for an app ID.
    pub fn find_existing_artwork(
        &self,
        user_id: &str,
        app_id: u32,
    ) -> Result<HashMap<ArtworkType, String>, SteamError> {
        let grid_dir = self.paths.grid_dir(user_id);
        let mut result = HashMap::new();

        let extensions = ["png", "jpg", "jpeg", "ico"];
        let art_types = [
            (ArtworkType::Grid, format!("{app_id}")),
            (ArtworkType::Hero, format!("{app_id}_hero")),
            (ArtworkType::Logo, format!("{app_id}_logo")),
            (ArtworkType::Icon, format!("{app_id}_icon")),
            (ArtworkType::Portrait, format!("{app_id}p")),
        ];

        for (art_type, base_name) in &art_types {
            for ext in &extensions {
                let path = grid_dir.join(format!("{base_name}.{ext}"));
                if path.exists() {
                    result.insert(*art_type, path.to_string_lossy().into_owned());
                    break;
                }
            }
        }

        Ok(result)
    }

    /// Saves artwork data to the appropriate path.
    pub fn save_artwork(
        &self,
        user_id: &str,
        app_id: u32,
        art_type: ArtworkType,
        data: &[u8],
        ext: &str,
    ) -> Result<(), SteamError> {
        self.ensure_grid_dir(user_id)?;

        let ext = ext.trim_start_matches('.');
        let ext = if ext.is_empty() { "png" } else { ext };

        let path = self.paths.artwork_path(user_id, app_id, art_type, ext);
        fs::write(&path, data)
            .map_err(|e| SteamError::Io(format!("failed to write artwork: {e}")))?;
        Ok(())
    }

    /// Removes all artwork for an app ID.
    pub fn delete_artwork(&self, user_id: &str, app_id: u32) -> Result<(), SteamError> {
        let existing = self.find_existing_artwork(user_id, app_id)?;
        for path in existing.values() {
            if let Err(e) = fs::remove_file(path)
                && e.kind() != std::io::ErrorKind::NotFound
            {
                return Err(SteamError::Io(format!("failed to remove {path}: {e}")));
            }
        }
        Ok(())
    }
}

/// Generates a Steam shortcut app ID from executable path and name.
///
/// Matches Steam's algorithm: `CRC32(exe + name) | 0x80000000 | 0x02000000`.
pub fn generate_app_id(exe: &str, name: &str) -> u32 {
    let key = format!("{exe}{name}");
    let mut hasher = Hasher::new();
    hasher.update(key.as_bytes());
    let crc = hasher.finalize();
    (crc | 0x80000000) | 0x02000000
}

/// Converts a `ShortcutConfig` to a `ShortcutInfo`.
pub fn convert_to_shortcut_info(cfg: &ShortcutConfig) -> ShortcutInfo {
    ShortcutInfo {
        app_id: generate_app_id(&cfg.exe, &cfg.name),
        name: cfg.name.clone(),
        exe: cfg.exe.clone(),
        start_dir: cfg.start_dir.clone(),
        launch_options: cfg.launch_options.clone(),
        tags: cfg.tags.clone(),
        last_played: 0,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn generate_app_id_deterministic() {
        let id1 = generate_app_id("/usr/bin/game", "My Game");
        let id2 = generate_app_id("/usr/bin/game", "My Game");
        assert_eq!(id1, id2);
    }

    #[test]
    fn generate_app_id_high_bits_set() {
        let id = generate_app_id("/bin/test", "Test");
        // Must have bits 0x80000000 and 0x02000000 set
        assert_ne!(id & 0x80000000, 0);
        assert_ne!(id & 0x02000000, 0);
    }

    #[test]
    fn generate_app_id_different_inputs() {
        let id1 = generate_app_id("/bin/a", "Game A");
        let id2 = generate_app_id("/bin/b", "Game B");
        assert_ne!(id1, id2);
    }

    #[test]
    fn convert_to_shortcut_info_basic() {
        let cfg = ShortcutConfig {
            name: "Test".into(),
            exe: "/bin/test".into(),
            start_dir: "/home".into(),
            launch_options: "--fullscreen".into(),
            tags: vec!["RPG".into()],
            artwork: None,
        };
        let info = convert_to_shortcut_info(&cfg);
        assert_eq!(info.name, "Test");
        assert_eq!(info.exe, "/bin/test");
        assert_eq!(info.app_id, generate_app_id("/bin/test", "Test"));
        assert_eq!(info.tags, vec!["RPG"]);
        assert_eq!(info.last_played, 0);
    }
}
